qualtran/bloqs/block_encoding/sparse_matrix.py

```diff
@@ -0,0 +1,279 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import abc
+from functools import cached_property
+from typing import cast, Dict, Tuple
+
+import numpy as np
+from attrs import frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    BoundedQUInt,
+    QAny,
+    QBit,
+    QUInt,
+    Register,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran.bloqs.basic_gates import Rx, Swap
+from qualtran.bloqs.block_encoding import BlockEncoding
+from qualtran.bloqs.block_encoding.lcu_select_and_prepare import PrepareOracle
+from qualtran.bloqs.state_preparation.prepare_uniform_superposition import (
+    PrepareUniformSuperposition,
+)
+from qualtran.symbolics import SymbolicFloat, SymbolicInt
+
+
+@frozen
+class RowColumnOracle(Bloq):
+    r"""Oracle specifying the non-zero rows or columns of a sparse-access matrix.
+
+    In the reference, this is the interface of
+    $O_r : \ket{\ell}\ket{i} \mapsto \ket{r(i, \ell)}\ket{i}$, and of
+    $O_c : \ket{\ell}\ket{j} \mapsto \ket{c(j, \ell)}\ket{j}$.
+    Here, $r(i, \ell)$ and $c(j, \ell)$ give the $\ell$-th nonzero entry in the $i$-th row
+    and $j$-th column of the matrix, respectively.
+
+    Registers:
+        l: As input, index specifying the `l`-th non-zero entry to find in row / column `i`.
+           As output, position of the `l`-th non-zero entry in row / column `i`.
+        i: The row / column index.
+
+    References:
+        [Lecture Notes on Quantum Algorithms for Scientific Computation](
+        https://arxiv.org/pdf/2201.08309). Lin Lin (2022). Ch. 6.5.
```

Comment:
extreme nit: can you link to the arxiv landing page rather than the pdf. And you can keep the link on one line (url and title) even if it goes too long

Reply:
Appreciated!